hazelcast/src/test/java/com/hazelcast/client/bluegreen/ClientClusterDiscoveryServiceTest.java

```diff
@@ -17,138 +17,248 @@
+        assertEquals(arrayList.get(3), searchAndGetNextCluster(3, discoveryService));
 
-        assertEquals(10 * n, count);
-        discoveryService.next();
+    }
 
+    private CandidateClusterContext searchAndGetNextCluster(int candidateIndex,
+                                                            ClusterDiscoveryService discoveryService) {
+        assert candidateIndex > 0;
+
+        final List<CandidateClusterContext> nextCluster = new ArrayList<>();
+        discoveryService.tryNextCluster((current, next) -> {
+            boolean foundNext = ++visitedIndex.value == candidateIndex - 1;
+            if (foundNext) {
+                nextCluster.add(next);
+            }
+            return foundNext;
+        });
+        return nextCluster.get(0);
     }
 
-    @Test(expected = NoSuchElementException.class)
-    public void test_behaviourWhenBlueGreenIsOff() {
-        ArrayList<CandidateClusterContext> arrayList = new ArrayList<CandidateClusterContext>();
-        CandidateClusterContext candidateClusterContext = createContext();
-        arrayList.add(candidateClusterContext);
+    @Test
+    public void test_n_iterations() {
+        ArrayList<CandidateClusterContext> arrayList = new ArrayList<>();
+        int n = 15;
+
+        for (int i = 0; i < 10; i++) {
+            arrayList.add(createContext(i));
+        }
+        ClusterDiscoveryService discoveryService = new ClusterDiscoveryService(arrayList, n, lifecycleService);
 
-        ClientDiscoveryService discoveryService = new ClientDiscoveryService(1, arrayList);
+        MutableInteger count = new MutableInteger();
+        discoveryService.tryNextCluster((o, o2) -> {
+            count.value++;
+            return false;
+        });
 
-        assertEquals(candidateClusterContext, discoveryService.next());
-        assertFalse(discoveryService.hasNext());
-        discoveryService.next();
+        assertEquals(10 * n, count.value);
 
     }
 
     @Test
     public void testCurrentAndNextReturnsCorrect() {
-        ArrayList<CandidateClusterContext> arrayList = new ArrayList<CandidateClusterContext>();
+        ArrayList<CandidateClusterContext> arrayList = new ArrayList<>();
 
-        CandidateClusterContext first = createContext();
+        CandidateClusterContext first = createContext(1);
         arrayList.add(first);
-        CandidateClusterContext second = createContext();
+        CandidateClusterContext second = createContext(2);
         arrayList.add(second);
-        ClientDiscoveryService discoveryService = new ClientDiscoveryService(10, arrayList);
+
+        ClusterDiscoveryService discoveryService = new ClusterDiscoveryService(arrayList, 10, lifecycleService);
 
         assertEquals(first, discoveryService.current());
-        discoveryService.resetSearch();
-        assertTrue(discoveryService.hasNext());
 
-        assertEquals(second, discoveryService.next());
+        assertEquals(second, searchAndGetNextCluster(1, discoveryService));
```

Comment:
I did not understand `searchAndGetNextCluster ` . Why are we giving the index again ? Wouldn't it be simpler to replace this method with something like: 
```
 private CandidateClusterContext getNextCluster(ClusterDiscoveryService discoveryService) {
        AtomicReference<CandidateClusterContext> currentCandidate = new AtomicReference<>();
        discoveryService.tryNextCluster((current, next) -> {
            currentCandidate.set(next);
            return true;
        });
        return currentCandidate.get();
    }
```

Reply:
done.